Alignment/APEEstimation/plugins/createDefaultApeTree.cc

```diff
@@ -0,0 +1,474 @@
+//
+//
+
+
+// system include files
+#include <memory>
+#include <fstream>
+#include <sstream>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+
+#include "CLHEP/Matrix/SymMatrix.h"
+
+#include "CondFormats/AlignmentRecord/interface/TrackerAlignmentErrorExtendedRcd.h"
+#include "CondFormats/Alignment/interface/AlignmentErrorsExtended.h"
+
+//...............
+#include "Alignment/APEEstimation/interface/TrackerSectorStruct.h"
+#include "Alignment/APEEstimation/interface/ReducedTrackerTreeVariables.h"
+
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "DataFormats/SiStripDetId/interface/StripSubdetector.h"
+#include "DataFormats/SiStripDetId/interface/SiStripDetId.h"
+#include "DataFormats/SiPixelCluster/interface/SiPixelCluster.h"
+#include "DataFormats/SiStripCluster/interface/SiStripCluster.h"
+
+
+#include "TString.h"
+#include "TFile.h"
+#include "TDirectory.h"
+#include "TTree.h"
+#include "TMath.h"
+//
+// class declaration
+//
+
+class DefaultApeTree : public edm::one::EDAnalyzer<> {
+  public:
+    explicit DefaultApeTree(const edm::ParameterSet&);
+    ~DefaultApeTree() override;
+  
+  
+  private:
+    void beginJob() override ;
+    void analyze(const edm::Event&, const edm::EventSetup&) override;
+    void endJob() override ;
+    
+    void sectorBuilder();
+    bool checkIntervalsForSectors(const unsigned int sectorCounter, const std::vector<double>&)const;
+    bool checkModuleIds(const unsigned int, const std::vector<unsigned int>&)const;
+    bool checkModuleBools(const bool, const std::vector<unsigned int>&)const;
+    bool checkModuleDirections(const int, const std::vector<int>&)const;
+    bool checkModulePositions(const float, const std::vector<double>&)const;
+    
+    // ----------member data ---------------------------
+    const edm::ParameterSet parameterSet_;
+    
+    std::map<unsigned int, TrackerSectorStruct> m_tkSector_;
+    std::map<unsigned int, ReducedTrackerTreeVariables> m_tkTreeVar_;
+};              
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+DefaultApeTree::DefaultApeTree(const edm::ParameterSet& iConfig):
+parameterSet_(iConfig)
+{
+}
+
+
+DefaultApeTree::~DefaultApeTree()
+{
+}
+
+
+//
+// member functions
+//
+void
+DefaultApeTree::sectorBuilder()
+{
+  // Same procedure as in ApeEstimator.cc
+  TFile* tkTreeFile(TFile::Open((parameterSet_.getParameter<std::string>("TrackerTreeFile")).c_str()));
```

Comment:
@mteroerd I believe it is recommended to use the TFileService, but @Dr15Jones might correct me.
At least, if you would use it you would have to specify this in the declaration of the Analyzer class.

Reply:
in addition, use FileInPath here.

Reply:
I can't find any documentations on TFileService or FileInPath that really explain how to correctly implement them or why they are better in this context. Can you link me to that, please?
Neither https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEdmFileInPath nor https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideTFileService are helpful to me.